### ISSUE TITLE
[BUG] C# dev kit for xUnit not discovering tests

### ISSUE THREAD

**Issue by snowberry-clearwing** [microsoft/vscode-dotnettools#307]
### Describe the Issue

I am using the C# dev kit extension for unit testing with xUnit. It was able to discover the tests that were already in the project when the addon was installed, however I cannot discover new tests. I am not sure if I am missing a setting somewhere, however I could not find any relevant documentation apart from https://code.visualstudio.com/docs/csharp/testing.

### Steps To Reproduce

1. Created a new xUnit test project.
2. Created a simple test.
3. Built the test project to ensure it compiles
4. Testing pane does not show new tests, even after hitting the 'Refresh Tests' icon.

### Expected Behavior

When I hit the 'Refresh Tests' icon in the Test explorer I would see updated tests.

### Environment Information

OS: WIndows 11
VS Code version:  1.80.2 (user setup)
Extension: v0.2.100

**Comment by snowberry-clearwing** [comment]
I figured why this was not working. The .sln file needed to be updated with the new test project. After I added this in Visual Studio, I could discover the tests in VS Code. Perhaps the aforementioned documentation could be checked to see if the setup instructions are correct? https://code.visualstudio.com/docs/csharp/testing.

**Comment by timheuer** [comment]
Thanks for the additional info @snowberry-clearwing. Just to confirm you already had an existing solution open, but you created the new test project and it didn't automatically add it to your open solution?

**Comment by snowberry-clearwing** [comment]
Yes, that is correct. I ran the 'dotnet add' command to ensure the project was added however the solution file still had to be fixed externally.

**Comment by timheuer** [comment]
Thanks @snowberry-clearwing -- so your steps were:
- Had an open solution
- Created a new test project <-- what path did you take for this? Started from command pallette with new project? And not from the existing solution view right?
- used `dotnet add` -- what command did you run here? `dotnet sln add`?

**Comment by ivanjx** [comment]
`dotnet clean` will immediately updates the tests pane for me.

**Comment by anuraj** [comment]
I am also facing the same issue. Just installed C# DevKit - I cloned this repo - https://github.com/anuraj/MinimalApi - In the editor I couldn't find any option to execute the tests.

![image](https://github.com/microsoft/vscode-dotnettools/assets/276411/d1fcff05-6f8a-49a8-b7f0-8181a93a1a2f)

I tried to remove the test project and added it again. I tried `dotnet sln remove` and `dotnet sln add` commands. But still VS Code is not showing any test options.

**Comment by timheuer** [comment]
Hi @anuraj - in VS Code there is a "Test Explorer" area as seen here where the tests show up.  After a build the tests will be discovered here:

![image](https://github.com/microsoft/vscode-dotnettools/assets/4821/9176f4df-7f91-4451-8d5d-396bbb6a0455)

**Comment by nwrkbiz** [comment]
> Hi @anuraj - in VS Code there is a "Test Explorer" area as seen here where the tests show up. After a build the tests will be discovered here:
> 
> ![image](https://user-images.githubusercontent.com/4821/273645061-9176f4df-7f91-4451-8d5d-396bbb6a0455.png)

I have the same issue, the tests do not show up within this view.

**Comment by wgrs** [comment]
This is still broken. None of my xunit tests are being discovered. Rebuilt. Re added to sln. Re installed vscode 1.84. Re installed dev kit. Still broken

**Comment by wgrs** [comment]
> > This is still broken. None of my xunit tests are being discovered. Rebuilt. Re added to sln. Re installed vscode 1.84. Re installed dev kit. Still broken
> 
> @wgrs is this a repo perhaps we can look at? What target framework versions are the tests/projects?

It's targeting 8. Calling dotnet test blows up with a missing testhost runtime json file error. Its a brand new install on Ubuntu 23.10. Should  be easy to replicate.

**Comment by timheuer** [comment]
> > > This is still broken. None of my xunit tests are being discovered. Rebuilt. Re added to sln. Re installed vscode 1.84. Re installed dev kit. Still broken
> > 
> > 
> > @wgrs is this a repo perhaps we can look at? What target framework versions are the tests/projects?
> 
> It's targeting 8. Calling dotnet test blows up with a missing testhost runtime json file error. Its a brand new install on Ubuntu 23.10. Should be easy to replicate.

Ok, the blow up on the CLI alone could be a root cause here (stating the obvious).  How did you install on Ubuntu? apt-get? Can you put the error on the dotnet-test issue here as well?

**Comment by wgrs** [comment]
Yes APT. Yes, I reported on vscode ghost issues.

**Comment by wgrs** [comment]
This is the error being produced when I try and run dotnet test on Ubuntu. I have checked and the reference to the file below is missing.
`Testhost process for source(s) '/home/xxx/Dev/Projects/FSBO/Tests/bin/Debug/net8.0/Tests.dll' exited with error: The specified runtimeconfig.json [/usr/lib/dotnet/sdk/8.0.100/testhost-8.0.runtimeconfig.json] does not exist
. Please check the diagnostic logs for more information.`

**Comment by timheuer** [comment]
Hmm… @peterwald this seems more fundamental on the environment.

**Comment by gui-lisboa** [comment]
I faced the same problem having projects and .sln file under a src/ folder. Reverted C# Devkit to previous version and it's working again. (from 1.0.14 to 0.5.150)

**Comment by peterwald** [comment]
This seems to be an issue with the local `dotnet` installation or configuration. I have not been able to reproduce the issues above. After running a build, the tests appear and are runnable in the test explorer pane. Closing due to lack of actionable details. Happy to re-investigate if more details come to light.

**Comment by earloc** [comment]
@peterwald, Closing this as "not reproducable" probably won´t help those people facing the issue (like me today).

I had tests showing up in one solution, but not in another.

working solution:
  - assembly with TargetFramework: `net6.0`
  - test-assembly with TargetFramework: `net8.0`
 
non-working solution:
  - assembly with TargetFramework: `net6.0`
  - tests-assembly with TargetFramework: `net7.0`
  
Just got hands on a fresh dev-machine (MacOS Sonoma@M3 Max) and double checked everything against the former rig (Windows 11@Intel).

The former (windows-)rig (where both solutions were successfully discover tests in test-explorer) has
- `net6.x`
- `net7.x`
- `net8.x`

installed.

The new (MacOS)-machine only has
- `net6.x`
- `net8.x`

installed.

I then just modified the test-project to target `net8.0`and voila, tests are showing up again in the test-explorer.

Seems as if the matching target-framework of the test-assembly has to be installed on the machine in order to let test-explorer discover tests.
(I did not try out to install `net7.x` to see if this also would´ve solved the issue, though)

Hope this might help anybody else facing a similar situation.

**Comment by westsiderz** [comment]
Facing the same issue here. I am using C# Dev Kit with Docker and Devcontainer. My machine is running Ubuntu but the project is opened inside the Docker container. I have `net7.x` The unit tests project is built with no issues. The Tests Explorer is not showing anything even if I refresh. As a matter of fact, it askes me to install additional plugin in order to be able to run tests.

**Comment by earloc** [comment]
@westsiderz Which version does your test-project is built against? For me, the issue was solved once the machine had the proper version installed (or the test-project actually compiled against a version that is present on the machine, respectively).

Probably also the case for a devcontainer?!

So either switch your test-projects target-framework or add the matching. net-sdk in your Container might help?

**Comment by cyates81** [comment]
I'm also having this issue. No matter what I try it says no test have been found even though the sln has the test project and dotnet test runs the tests properly. I am running in MacOS with all projects targeting 8.0.

**Comment by peterwald** [comment]
The following steps should work to get some sample xUnit tests working as a baseline for comparison.

Prerequisites
1. .NET 8 SDK is installed. Check version using `dotnet --list-sdks`. (this should work with most of the recent versions of the SDK as well, net 6 or net 7 for example).
2. VS Code is installed and the C# Dev Kit extension is also installed in VS Code (tested with version v1.3.6)

Steps
1. `mkdir mytests`
2. `cd mytests`
3. `dotnet new xunit`
4. `dotnet new sln`
5. `dotnet sln add .`
6. `dotnet test` - Tests should run and succeed
7. `code .` - Open VS Code in the test folder

After initialization is complete, the test explorer panel should show the sample test from this project.

**Comment by binair-io** [comment]
@peterwald Everything works up until step 7. I can run tests using the CLI, but they don't appear in the test explorer. Tried cleaning, rebuilding, dotnet test, closing and reopening window. 

- TargetFramework: net8.0
- Microsoft.NET.Test.Sdk: v17.7.2
- xunit: v2.6.0
- xunit.runner.visualstudio: v2.5.6
- c# dev kit: v1.3.10

**Comment by nestorbfgo** [comment]
> Thanks @snowberry-clearwing -- so your steps were:
> 
> * Had an open solution
> * Created a new test project <-- what path did you take for this? Started from command pallette with new project? And not from the existing solution view right?
> * used `dotnet add` -- what command did you run here? `dotnet sln add`?

Mabe late but this works

**Comment by omarmciver** [comment]
> `dotnet clean` immediately updates the tests pane for me. for convenience, i do `dotnet watch clean`.

I find that I have to do an `rm -rf '**/bin/** && rm -rf **/obj/**` on occasions to get the C# DevKit test explorer to detect properly. The `dotnet clean` alone doesn't do it, which is weird. Might be Devcontainer related (building in it and out of it and causing file permission confusion).

**Comment by Sharaf0** [comment]
Unfortunately I still have the same problem on mac, I tried every solution provided.

**Comment by Farazzaidi22** [comment]
There are a lot of mix response in this chat. What worked for me is a mix a couple of answers I saw. 

1. Run `dotnet clean`
2. Run `dotnet build`

Then tests started to appear

![image](https://github.com/user-attachments/assets/92bf566b-24ee-4b5a-b5af-49edae0a841d)

**Comment by chrisdel101** [comment]
I also had this. Test explorer has been working and stopped all of a sudden. Build and clean did nothing. Restarting vscode did nothing. Even typing Tests in the command palette returned no results. There were no errors in the output panel. There was nothing left to try. 

Two days later and a few restarts later the tests show up. So I cannot reproduce the issue either.

**Comment by nwrkbiz** [comment]
For me the test project at least need to be net7.0 for the test runner to detect the unittests.

**Comment by amakhno** [comment]
Cleaning the test store folder helped. You can find it in the Output -> C# Dev Kit tab

```
Created Test Controller
Log verbosity: minimal
platform: win32 arch: x64
Using vstest from dotnet sdk in [C:\Program Files\dotnet\sdk\8.0.307].
Initialized Test Explorer Server [21964]
Test Store Folder: c:\Users\<Username>\AppData\Roaming\Code\User\workspaceStorage\8d49cae0d505b6684c6ba6a076c9a18c\ms-dotnettools.csdevkit
```

**Comment by mattmc3** [comment]
I had this problem with tests in a project with submodules. Those submodules had their own solution files. When I looked at the output window under "C# Dev Kit - Test Explorer" it let me know it "Cannot read runsettings file at .../.runsettings_local" so I made that files. When I looked under the output for "C# Dev Kit", it told me "More than one solution files in the workspace. Waiting for the developer to make a choice.", so I chose ".NET: Open Solution" from the command palette. Doing both of those things solved my issue.

The Google results for this aren't very helpful, so I'm posting this in case someone else has this problem and hopefully finds this. Not sure why vscode doesn't prompt to pick a solution, or just default to the root one.
